pkg/cmd/server/handlers/authentication.go

```diff
@@ -9,20 +9,29 @@ import (
 	apirequest "k8s.io/apiserver/pkg/endpoints/request"
 )
 
+const (
+	AuditAuthnUser  = "audit-authn-user"
+	AuditAuthnError = "audit-authn-error"
+)
+
 // AuthenticationHandlerFilter creates a filter object that will enforce authentication directly
 func AuthenticationHandlerFilter(handler http.Handler, authenticator authenticator.Request, contextMapper apirequest.RequestContextMapper) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		ctx, ok := contextMapper.Get(req)
+		if !ok {
+			http.Error(w, "Unable to find request context", http.StatusInternalServerError)
+			return
+		}
+
 		user, ok, err := authenticator.AuthenticateRequest(req)
 		if err != nil || !ok {
+			ctx = apirequest.WithValue(ctx, AuditAuthnUser, user)
+			ctx = apirequest.WithValue(ctx, AuditAuthnError, err)
```

Comment:
@smarterclayton I've tried this approach to get the user and error and return that in the fallback audit logs, but it doesn't work :/ @sttts proposed we change `AuthenticateRequest` signature by extending its return type with additional `[]AuthenticationResult` which will provide the information about user and method, with possibility to extend that in the future. We'd upstream this into 1.8, obviously. The only problem is that it might require a bit more work, and we're just closing 3.6. The other approach is to duplicate the logic responsible for getting user information in the fallback audit, and for token requests we'd just say it's that. wdyt?